clients/spark/core/src/test/scala/io/treeverse/clients/ApiClientSpec.scala

```diff
@@ -32,17 +38,136 @@ class ApiClientSpec extends AnyFunSpec with Matchers {
       it("should translate any protocol to abfs") {
-        translate("https://account.example.net/container/path/to/blob") should be("abfs://container@account.dfs.core.windows.net/path/to/blob")
-        translate("ex://account.example.net/container/path/to/blob") should be("abfs://container@account.dfs.core.windows.net/path/to/blob")
+        translate("https://account.example.net/container/path/to/blob") should be(
+          "abfs://container@account.dfs.core.windows.net/path/to/blob"
+        )
+        translate("ex://account.example.net/container/path/to/blob") should be(
+          "abfs://container@account.dfs.core.windows.net/path/to/blob"
+        )
       }
 
       it("should handle empty paths") {
         // TODO(lynn): Trailing slashes might well be incorrect here.
-        translate("https://account.example.net/container/") should be("abfs://container@account.dfs.core.windows.net/")
-        translate("https://account.example.net/container") should be("abfs://container@account.dfs.core.windows.net/")
+        translate("https://account.example.net/container/") should be(
+          "abfs://container@account.dfs.core.windows.net/"
+        )
+        translate("https://account.example.net/container") should be(
+          "abfs://container@account.dfs.core.windows.net/"
+        )
       }
     }
   }
 }
+
+class RequestRetryWrapperSpec extends AnyFunSpec with Matchers with BeforeAndAfterEach {
+  val SomeValue = "some value"
+  val DefaultMaxNumRetries = 5
+  val DefaultMaxNumAttempts = 6
+
+  var retryWrapper = spy(new RequestRetryWrapper)
+  var dummyMethodInvoker = spy(new DummyMethodInvoker)
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    retryWrapper = spy(new RequestRetryWrapper)
+    assert(retryWrapper.numRetries == DefaultMaxNumRetries)
+    dummyMethodInvoker = spy(new DummyMethodInvoker)
+  }
+
+  class DummyMethodInvoker {
+    @throws[Exception]
+    def someMethod(): String = {
+      SomeValue
+    }
+
+    def createCheckedSupplierForSomeMethod(): dev.failsafe.function.CheckedSupplier[String] = {
+      new dev.failsafe.function.CheckedSupplier[String]() {
+        def get(): String = someMethod()
+      }
+    }
+  }
+
+  describe("wrapWithRetry") {
+    // In case all retries fail, the last result or exception are returned or thrown
+    // https://failsafe.dev/faqs/#how-to-i-throw-an-exception-when-retries-are-exceeded
+    it(
+      "should throw the a FailSafeException exception in case all retries failed and last exception is checked"
+    ) {
+      // prepare
+      doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[ApiException])
+        .when(dummyMethodInvoker)
+        .someMethod()
+
+      assertThrows[FailsafeException] {
+        retryWrapper.wrapWithRetry(dummyMethodInvoker.createCheckedSupplierForSomeMethod())
+      }
```

Comment:
IIUC this says that this retry loses all information about which checked exception was actually thrown!  It threw `ApiException`, and we received `FailsafeException`.  But if that happens, how do we debug a failure that gave us `ApiException`s?